### DISCUSSION TITLE
Multi-instance stores rooted to specific component instances?

### DISCUSSION THREAD

**Discussion by Infeligo** [vuejs/pinia#840] [Ideas]
I'll try to explain the idea by describing my use case. Maybe, there is already a better solution for it.

Imagine you need to implement something like a datagrid - a searchable, sortable, filterable and pageable table of some data. This is a complex component composed of many sub-components. Naturally, you want to separate the state management part into... a store. This would keep the current search and sort criteria, found results and maybe some state related to the appearance and behavior of the datagrid parts. It would also have actions to call search API based on the current state. Such store would allow all sub-components to access the data and to modify it (e.g. trigger search) in a controlled and uniform way.

The problem is that all the store libraries like Vuex and Pinia provide global central stores, which are tied to the application root. However, there are valid use cases when several datagrids are displayed on a single page (e.g. some kind data comparison). The state of each datagrid component tree should be independent from others and its lifecycle should be bound to the lifecycle of its root component.

Can this be achieved with Pinia?

**Comment by joshkrz** [comment]
I know that you can spin up [multiple instances of Vuex modules](https://vuex.vuejs.org/guide/modules.html#module-reuse). I'm wondering if this can be achived in Pinia to solve the exact situation your describing.

**Comment by ccqgithub** [comment]
> I'll try to explain the idea by describing my use case. Maybe, there is already a better solution for it.
> 
> Imagine you need to implement something like a datagrid - a searchable, sortable, filterable and pageable table of some data. This is a complex component composed of many sub-components. Naturally, you want to separate the state management part into... a store. This would keep the current search and sort criteria, found results and maybe some state related to the appearance and behavior of the datagrid parts. It would also have actions to call search API based on the current state. Such store would allow all sub-components to access the data and to modify it (e.g. trigger search) in a controlled and uniform way.
> 
> The problem is that all the store libraries like Vuex and Pinia provide global central stores, which are tied to the application root. However, there are valid use cases when several datagrids are displayed on a single page (e.g. some kind data comparison). The state of each datagrid component tree should be independent from others and its lifecycle should be bound to the lifecycle of its root component.
> 
> Can this be achieved with Pinia?

maybe this can help you? [pinia-di](https://github.com/ccqgithub/pinia-di).

**Comment by markbaydoun** [comment]
Don't know if this still applies, but I found a way to do it like you say

`useTableStore.ts`
```typescript
export const useTableStore = (tableId: string) => defineStore(`tables/${tableId}`, {
  state: () => ({
    foo: {},
    bar: {} 
  }),
  getters: ...,
  actions: ...,
})(); // <- this is the key right here, by initialising it here
```

#

Then in other component or components:
```typescript
const tableId = props.table.id // Just as an example

const tableStore = useTableStore(tableId)
```

Assuming the tableIds are different for each table, they will use a new instance of this store, so they won't share state unless the tableId is the same

**Reply by posva** [reply]
This would be the way but you also need to not redefine each time the store. So cache the returned values of defineStore:

```ts
const definedStores = new Map<string, ReturnType<typeof defineTableStore>>()
export const useTableStore = (tableId: string) => {
  if (!definedStores.has(tableId)) {
    definedStores.set(
      tableId,
      defineTableStore(tableId)
    )
  }

  return definedStores.get(tableId)!
}

function defineTableStore<Id extends string>(tableId: Id) {
  return defineStore(`tables/${tableId}`, {
    /* ... */
  })
}
```

**Reply by markbaydoun** [reply]
This works as well, but I have lost type inference of the state, getters, etc for some reason(probably because I don't know enough about it). Doing it the other way only showed one instance of the store in Vue devtools and everything worked as I expected, even though I used it in multiple places, are you sure I was instantiating a new one every time I called useTableStore() ?

**Reply by posva** [reply]
It does need a few changes to support typing so I updated the comment. I think this could be added to docs. Not sure where though.

**Reply by markbaydoun** [reply]
That'd be great, I think I won't be the only one to come across this question.

So the way I was doing it was actually making multiple instances of the same store right?

**Reply by bakodc** [reply]
Thanks. That works perfect.  I use it to store/persist the listview state of multiple component instances.
The Vue tools show one store instance per component. I adjusted your code a bit, so it actually returns a reference to the store instead of a function.

```ts
import { defineStore } from "pinia";
export interface IListViewState {
  sortColumn: string;
  sortOrder: SortOrder;
  filter: ColumnFilter[];
  page: number;
}

const definedStores = new Map<string, ReturnType<typeof defineListViewStore>>();
const listviewStoreFactory = (listViewId: string) => {
  if (!definedStores.has(listViewId)) {
    definedStores.set(listViewId, defineListViewStore(listViewId));
  }

  return definedStores.get(listViewId) as ReturnType<
    typeof defineListViewStore
  >;
};

function defineListViewStore<Id extends string>(listViewId: Id) {
  return defineStore(`listView/${listViewId}`, {
    state: (): IListViewState => {
      return {
        sortColumn: "",
        sortOrder: "ASC",
        filter: [],
        page: 0,
      };
    },
 /* ...*/
  });
}
export const useListviewStore = (listViewId: string) => {
  return listviewStoreFactory(listViewId)();
};
```

**Reply by radusuciu** [reply]
Great solution! Is it possible to make it work with HMR or do we give that up?

```typescript
if (import.meta.hot) {
    import.meta.hot.accept(acceptHMRUpdate(useListviewStore, import.meta.hot))
}
```

This throws a type error:

```
Argument of type '(listViewId: string) => Store<`listView/${string}`, IListViewState, {}, {}>' is not assignable to parameter of type 'StoreDefinition<string, StateTree, _GettersTree<StateTree>, _ActionsTree>'.
  Property '$id' is missing in type '(listViewId: string) => Store<`listView/${string}`, IListViewState, {}, {}>' but required in type 'StoreDefinition<string, StateTree, _GettersTree<StateTree>, _ActionsTree>'.ts(2345)
pinia.d.ts(622, 5): '$id' is declared here.
```

This makes sense since with the code above we're not passing a `listViewId` to `useListViewStore`, but I'm not sure where to get `listViewId` in this context.

Note I replaced the `SortColumn` and `ColumnFilter` types with `string`, but I don't think this impacts things as I hit the same error with a different implementation.

**Reply by 3zzy** [reply]
@posva I still can't find anything in the docs. Should I just use this or there're a newer/preferred way or instantiating multiple stores now? Thanks!

**Reply by danielnaca** [reply]
Thank you so much, I've been stuck on this for a few days. Mirroring what 3zzy was saying,  @posva can you recommend which of the above code blocks to use.

Would it work if I rewrote it in JS instead of TS?

**Reply by denjaland** [reply]
This looks like a great solution.  My only question is: if the user is browsing a lot of pages, are we not risking the definedStores to continue growing.  I was looking at a way to detect unused stores so they can be cleaned up too.

**Reply by miro-ku** [reply]
I've written a helper function, that makes this easier to do not duplicate code in few stores, if needed, and more clean code in store definition file

I'm quite new to typescript, so appreciate any suggestions to improve this. (For example to make this function throw TS error, if argument passed not a function, that returns `StoreDefinition`)

```typescript
export default function makeSeparatedStore<
  T extends (storeKey: string) => any,
  K extends T extends (storeKey: string) => infer StoreDef ? StoreDef : never,
>(defineStore: T) {
  const definedStores = new Map<string, K>();

  return (
    storeKey: string,
  ): ReturnType<K> => {
    if (!definedStores.has(storeKey)) {
      definedStores.set(storeKey, defineStore(storeKey));
    }

    // @ts-expect-error
    return definedStores.get(storeKey)();
  };
}
```

Usage:

```typescript
export const useTableStore = makeSeparatedStore((key: string) =>
  defineStore(`table/${key}`, {
    state: (): StateType => ({
      /* ... */
    }),
    getters: {
      /* ... */
    },
    actions: {
      /* ... */
    },
  }),
);
```

**Reply by 3zzy** [reply]
@posva How do I pass the ID when when using `mapStores` helper in options API? I'm trying to use an `injected` ID eg `...mapStores(useFormEditor(this.myid))` but it doesn't seem to work. Would really appreciate any help.

**Reply by dgioulakis** [reply]
@3zzy You should be able to use provide/inject
e.g.
```
    const loadingSheetId = inject<string>("loadingSheetId")!;
    const loadingStore = useLoadingSheetStore(loadingSheetId); // << calls makeSeparatedStore()
```
Create a component which provides your unique store key. This seems to work quite well for me, but I haven't tested thoroughly.

**Reply by sabifa** [reply]
This discussion is pretty dated already, is this still the recommended way? @posva 
You stated that it could be added to the docs but it seems like that never happened. Is there a reason to it?

**Reply by MickL** [reply]
Works fine for me:

```ts
export const useProductStore = (productId: number) =>
  defineStore(`product-${productId}`, () => {
      console.log('create product store for ' + productId);
  };
})();
```

**Reply by miro-ku** [reply]
@MickL You should use `productId` in store name

**Reply by MickL** [reply]
You are so fast to reply :) I just found it out as well and updated my answer. Works totally fine, stores that are created once are cached as well.

I just wonder about garbage collection ... By time hundreds of stores could be created and will never be destroyed?

**Reply by miro-ku** [reply]
Yea, you should definitely determine the moment, when the store isn't needed anymore, and destroy it manually

**Reply by em3link** [reply]
For anyone wondering, this doesn't seem to be documented, but you can destroy a store created dynamically using `storeInstance.$dispose()`.

In my case, I have a collection of items (with computed properties) which I need to access in several components, so I created a dynamic store `Map()` like in the examples above.
I also keep track of an array of `itemIds` in a global store state.
Then I can remove an item from the collection by querying it from the `Map` and using `$dispose()` first, then deleting it from the `Map`, and finally splicing the id from the `itemIds` array.

Below a stripped-down sample from my code.
```typescript
const itemsStores = new Map<number, ReturnType<typeof defineItemStore>>()

export const useItemStore = (itemId: number) => {
  return itemsStores.get(itemId)
}

function defineItemStore(item: Item) {
  return defineStore(`items/${item.id}`, () => {
    const id = ref(item.id)
    const name = ref(item.name)
    const quantity = ref(item.quantity)
    const price = ref(item.price)

    // Example of a computed value I need
    const costTotal = computed(() => price.value * quantity.value)

    return {
      id,
      name,
      quantity,
      price,
      costTotal
    }
  })()
}

export const useGlobalStore = defineStore('global', () => {
    const itemIds = ref<number[]>([])
    const _nextId = ref<number>(0)

    function addItem(name: string, price: number) {
      itemsStores.set(
        _nextId.value,
        defineItemStore({
          id: _nextId.value,
          name,
          quantity: 1,
          price,
        })
      )
      itemIds.value.push(_nextId.value)
      _nextId.value++
    }

    function removeItem(index: number) {
      deviceItemsStores.get(itemIds.value[index])!.$dispose()
      deviceItemsStores.delete(itemIds.value[index])
      itemIds.value.splice(index, 1)
    }

    return {
      itemIds,
      _nextId,
      addItem,
      removeItem
    }
  })()
}
```

**Comment by aghArdeshir** [comment]
I'm not sure if I'm the only one who understood the problem differently or not.
But I had the same issue, and none of the replies really worked fr me.
What worked for me in the end was to `externalize` pinia using webpack in the library and specifying it only as a peerDependency. So there was no pinia bundled in the final build.

**Reply by pcoterecollective** [reply]
Can you give more details please?

**Reply by aghArdeshir** [reply]
I really don't remember what was the challenge I had at that time :sweat_smile: 

This is the comment I left at the time 

![image](https://github.com/vuejs/pinia/assets/5755214/afeb7c12-2f79-4492-95f5-63208d7243ff)

We have a root project (normal nuxt + vue + vite app). It included some child projects: the master app mounted the children (children each export a Vue component). Both the master project and the child project had their own stores in vuex, and (unfortunately) they had access to each other's stores ,they could read and write (with Vuex this was possible, because each Vue app (the master project) has one vuex instance that each component can get access to).

But with Pinia this was not possible directly, and I wanted to keep the code that allows reading and writing the stores (master <---> child). If I remember correctly, after investigations, I found out if two projects have two different `pinia` npm dependecies installed, then the runtime pinia (`createPinia()`) are different. So I externalized it in the child project, so in the final bundle, there is only one pinia)

**Comment by etranger** [comment]
Apparently, you can use internal component property `this._uid` as part of store identifier
